File: src/main/scala/ru/tinkoff/gatling/amqp/request/PublishDslBuilderExchange.scala
package ru.tinkoff.gatling.amqp.request

import io.gatling.core.config.GatlingConfiguration
import io.gatling.core.session.Expression

case class PublishDslBuilderExchange(
    requestName: Expression[String],
    configuration: GatlingConfiguration
) {
  def topicExchange(name: Expression[String], routingKey: Expression[String]):PublishDslBuilderMessage =
    destination(AmqpTopicExchange(name, routingKey = routingKey))

  def directExchange(name: Expression[String], routingKey: Expression[String]): PublishDslBuilderMessage =
    destination(AmqpDirectExchange(name, routingKey))

  def queueExchange(name: Expression[String]): PublishDslBuilderMessage = destination(AmqpQueueExchange(name))

  protected def destination(dest: AmqpExchange) = PublishDslBuilderMessage(requestName, dest, configuration)
}

File: src/main/java/ru/tinkoff/gatling/amqp/javaapi/request/PublishDslBuilderExchange.java
package ru.tinkoff.gatling.amqp.javaapi.request;

import static io.gatling.javaapi.core.internal.Expressions.*;
public class PublishDslBuilderExchange {

    private final ru.tinkoff.gatling.amqp.request.PublishDslBuilderExchange wrapped;

    public PublishDslBuilderExchange(ru.tinkoff.gatling.amqp.request.PublishDslBuilderExchange wrapped) {
        this.wrapped = wrapped;
    }

    public PublishDslBuilderMessage topicExchange(String name, String routingKey){
        return new PublishDslBuilderMessage(wrapped.topicExchange(toStringExpression(name),toStringExpression(routingKey)));
    }

    public PublishDslBuilderMessage directExchange(String name, String routingKey){
        return new PublishDslBuilderMessage(wrapped.directExchange(toStringExpression(name),toStringExpression(routingKey)));
    }

    public PublishDslBuilderMessage queueExchange(String name){
        return new PublishDslBuilderMessage(wrapped.queueExchange(toStringExpression(name)));
    }
}
